File: helix/src/main/java/glitch/helix/service/AnalyticsService.java
package glitch.helix.service;

import glitch.auth.objects.json.Credential;
import glitch.helix.GlitchHelix;
import glitch.helix.service.request.ExtensionAnalyticsRequest;
import glitch.helix.service.request.GameAnalyticsRequest;
import glitch.service.AbstractHttpService;

public class AnalyticsService extends AbstractHttpService {
    public AnalyticsService(GlitchHelix helix) {
        super(helix.getClient(), helix.getHttpClient());
    }

    public ExtensionAnalyticsRequest getExtensionAnalytics(Credential credential) {
        return new ExtensionAnalyticsRequest(http, credential);
    }

    // TODO
    public GameAnalyticsRequest getGameAnalytics() {
        throw new UnsupportedOperationException("Game Analytics is not available yet.");
    }
}

File: helix/src/main/kotlin/io/glitchlib/api/service/AnalyticsService.kt
package io.glitchlib.api.service

import io.glitchlib.GlitchClient
import io.glitchlib.api.internal.AbstractHelixService
import io.glitchlib.api.model.ExtensionReport
import io.glitchlib.api.model.GameReport
import io.glitchlib.api.service.request.ExtensionAnalyticsRequest
import io.glitchlib.api.service.request.GameAnalyticsRequest
import io.glitchlib.auth.Credential
import io.glitchlib.auth.Scope
import io.glitchlib.internal.http.bodySingle
import io.glitchlib.model.CursorList
import io.glitchlib.model.OrdinalList

class AnalyticsService(client: GlitchClient) : AbstractHelixService(client) {
    fun getExtensionAnalytics(credential: Credential, request: ExtensionAnalyticsRequest.() -> Unit = {}) =
        if (credential.scopeCheck(Scope.ANALYTICS_READ_EXTENSIONS))
            get<OrdinalList<ExtensionReport>>(
                "/analytics/extensions",
                ExtensionAnalyticsRequest(credential).apply(request)()
            ).bodySingle
        else scopeIsMissing(Scope.ANALYTICS_READ_EXTENSIONS)

    fun getGameAnalytics(credential: Credential, request: GameAnalyticsRequest.() -> Unit = {}) =
        if (credential.scopeCheck(Scope.ANALYTICS_READ_GAMES))
            get<CursorList<GameReport>>(
                "/analytics/extensions",
                GameAnalyticsRequest(credential).apply(request)()
            ).bodySingle
        else scopeIsMissing(Scope.ANALYTICS_READ_GAMES)
}
